python_ta/typecheck/base.py

```diff
@@ -21,12 +21,59 @@ def __init__(self, node_type: type, ast_node: Optional[NodeNG] = None):
         self.adj_list = []
         self.ast_node = ast_node
 
+    def __hash__(self):
+        return hash(repr(self))
+
     def __eq__(self, other: '_TNode'):
         if str(self.type) == str(other.type):
             return True
         else:
             return False
 
+    def __str__(self):
+        if self.parent and self.ast_node:
+            return f'TNode {self.ast_node.as_string()}: {self.type}, resolved to {self.parent.type}'
+        elif self.ast_node:
+            return f'TNode {self.ast_node.as_string()}: {self.type}'
+        else:
+            return f'TNode: {self.type}'
+
+    def get_nodes_in_reach(self):
+        open_nodes = [self]
+        visited_nodes = []
+        while open_nodes:
+            cur_open_node = open_nodes[0]
+            for e in cur_open_node.adj_list:
+                if e[0] not in visited_nodes and e[0] not in open_nodes:
+                    open_nodes.append(e[0])
+            visited_nodes.append(cur_open_node)
+            open_nodes.remove(cur_open_node)
+        return visited_nodes
+
+    def find_path(self, goal_node: '_TNode'):
```

Comment:
This part is unnecessary. Modify the existing code to store not just the "parent" but a "path to the parent" as well. If you do this well then you should be able to remove the hack of `self.parent is parent` check.